Rename CMake targets to avoid clashes when used as a subproject

ccawley2011 opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Added a user-facing option to provide zlib-compatible target aliases.
- Refactor
  - Public libraries renamed to zlib-ng (shared) and zlib-ng-static (static); naming, versioning, output names and DLL suffix handling standardized.
- Chores
  - Installation/export and generated config outputs updated to publish zlib-ng‑named libraries and symbols.
- Tests
  - Tests, benchmarks and fuzz targets updated to link against zlib‑ng variants with no behavior changes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ccawley2011 committed:
Rename CMake targets to avoid clashes when used as a subproject

ccawley2011 committed:
Provide zlib-ng::zlib aliases when used as a subproject

mtl1979 commented:
Half of the build directories are broken due to lack of updating... Not sure if this is even a good idea as any third-party build scripts will break too.

nmoinvaz commented:
This will break a lot of downstream projects that rely on the old CMake project names.

nmoinvaz commented:
This might be relevant:
https://github.com/zlib-ng/zlib-ng/issues/723

ccawley2011 committed:
Allow the previous target names to be used as aliases

ccawley2011 commented:
> Half of the build directories are broken due to lack of updating...

Fixed.

> This will break a lot of downstream projects that rely on the old CMake project names.

I've added optional aliases for the old names - projects can now use `set(ZLIB_ALIASES OFF)` to avoid the clash, while existing projects should be unaffected.

mtl1979 commented on test/CMakeLists.txt:
```diff
@@ -239,9 +239,9 @@ if(WITH_GTEST)
             # Link statically in order to test internal zlib-ng functions.
-            target_link_libraries(gtest_zlib zlibstatic)
+            target_link_libraries(gtest_zlib zlib-ng-static)
         else()
-            target_link_libraries(gtest_zlib zlib)
+            target_link_libraries(gtest_zlib zlib-ng)
         endif()
```
When building only static library, the default target name after this pull request is `zlib-ng`, as differentiating between the two build types is unnecessary. Using CMake aliases only works in third-party projects that include zlib-ng.

mtl1979 commented:
> > This will break a lot of downstream projects that rely on the old CMake project names.
> 
> I've added optional aliases for the old names - projects can now use `set(ZLIB_ALIASES OFF)` to avoid the clash, while existing projects should be unaffected.

This is a good idea and cause less trouble, especially with older CMake versions that can't handle duplicate target names. Eventually, in future major release, we might want to drop the support for short target names as recent releases of stock zlib already support the namespaced target names.

nmoinvaz commented:
What about the GH issue where the customize the entire name?

mtl1979 commented:
> What about the GH issue where the customize the entire name?

Customizing the entire name might cause own issues... It might be better for a zlib-ng fork instead of the main repository just to keep things wandering too much and becoming maintaining nightmare.

nmoinvaz commented:
In minizip-ng I do support specifying a suffix to the project name.

mtl1979 commented:
> In minizip-ng I do support specifying a suffix to the project name.

When the aliases are disabled, nothing stops from supporting specifying a suffix for a target name, but it might warrant separate PR after this one is merged... Biggest issue with supporting completely different name for targets in zlib-ng is that the new name can clash with any random target in any project and then people come to us complaining things break.

Later CMake versions can already handle duplicate target names in subprojects, so this is mainly for backwards compatibility until we stop supporting old enough CMake versions.

nmoinvaz commented:
> Biggest issue with supporting completely different name for targets in zlib-ng is that the new name can clash with any random target in any project and then people come to us complaining things break.

We can’t fix stupid people.

My point in mentioning that other GH issue is that we should just implement that and these changes in this PR would not be needed.

mtl1979 commented:
> > Biggest issue with supporting completely different name for targets in zlib-ng is that the new name can clash with any random target in any project and then people come to us complaining things break.
> 
> We can’t fix stupid people.
> 
> My point in mentioning that other GH issue is that we should just implement that and these changes in this PR would not be needed.

We also shouldn't make things easier for stupid people... It's always about choosing what is the best way to implement things, sometimes it's not the one with smallest or least complicated change. I've been software designer and developer for over 40 years and I've had quite a lot of experience with choosing what is the best option of many choices... But in the end it's just one opinion when I'm not the project maintainer.

Dead2 commented:
1. I do prefer that we don't keep the name zlib/zlibstatic all over CMake, those should really only be for zlib-compat mode. So in that way I agree with this PR. ✔️
2. Having the `zlib-ng::zlib` style namespacing ✔️ 
3. I would have also liked to have seen support for adding custom aliases like #723 suggests ❌ 

What do you think about point 3 @ccawley2011, is this something you would want to implement as well?

Dead2 commented:
@ccawley2011 Would you please rebase this PR, so we can see that it still works after other PRs. And I'd also like your reply on the above question if you could.

Dead2 reviewed (approved):
This LGTM.  Perhaps someone else wants to make a PR to add CMake target aliases

mtl1979 commented:
> This LGTM. Perhaps someone else wants to make a PR to add CMake target aliases

Adding "target aliases" needs proper discussion or it will flood us with complains from people who try to use the feature without knowing the restrictions for target names.

In my professional opinion, we should only allow adding extra suffix to the library base name(s), as that should be enough for portable builds, which would be the biggest user of the feature.

ccawley2011 commented:
The use case in the linked issue suggests that this might be sufficient since it both renames the target to be different from standard zlib and adds `zlib-ng::zlib` and `zlib-ng::zlibstatic` as aliases, but it would be good if @joshtriplett is able to confirm if this is good enough or not.

mtl1979 commented:
For portable builds, we need to adjust `zlib-ng-config.cmake` or `zlib-config.cmake` so it doesn't reference `CMAKE_CURRENT_LIST_DIR`, and sets `*_INCLUDE_DIRS` to both `*_BINARY_DIR` and `*_SOURCE_DIR`, where `*` is either `zlib`or `zlib-ng`, in that order ... Then we can export a target, like `zlib-portable` or `zlib-ng-portable`, that doesn't exist in normal build. Each target can be associated with a component that can be used with `find_package` to differentiate types of builds. 

A lot of the information is scattered through several sections of various books about CMake, with very little information in actual CMake documentation except notes about deprecated features.

This will make sure `find_package` finds the portable build before the regular build.
